### PR TITLE
Add privacy debug log

### PR DESCRIPTION
**Pull request by And1sS**

### CODE REVIEW

**Comment by snahornyi** [comment]
Tests in https://github.com/prebid/prebid-server-java/pull/1577

File: src/main/java/org/prebid/server/privacy/PrivacyExtractor.java [hunk 1]
```diff
@@ -89,6 +95,19 @@ private static Privacy toValidPrivacy(String gdpr,
         return Privacy.of(validGdpr, validConsent, validCcpa, validCoppa);
     }
 
+    private Privacy extractOriginPrivacy(Regs regs, User user, List<String> errors) {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/privacy/PrivacyExtractor.java:98]: Can be static?

File: src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        PrivacyCcpaDebug ccpa;
+
+        @JsonInclude(value = JsonInclude.Include.NON_EMPTY)
+        PrivacyCoppaDebug coppa;
+    }
+
+    @Value(staticConstructor = "of")
+    private static class PrivacyTcfDebug {
+
+        public static PrivacyTcfDebug of(String gdpr, String tcfConsentString) {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java:97]: Let's declare methods after fields.

File: src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+     */
+    private static boolean shouldMaskUser(User user) {
+        if (user == null) {
+            return false;
+        }
+        if (user.getId() != null || user.getBuyeruid() != null) {
+            return true;
+        }
+        final ExtUser extUser = user.getExt();
+        return extUser != null && (CollectionUtils.isNotEmpty(extUser.getEids()));
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java:204]: Redundant braces.

File: src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+            maskFunctions.add((device, deviceBuilder) -> deviceBuilder.ipv6(ipAddressHelper
+                    .anonymizeIpv6(device.getIpv6())));
+            logMessages.add(String.format("Device IPs were masked in request to bidder according to %s policy.",
+                    enforcementPolicy));
+            return this;
+        }
+
+        public DeviceAnonymizer maskGeoCoordinates() {
+            maskFunctions.add((device, deviceBuilder) ->
+                    deviceBuilder.geo(PrivacyAnonymizationService.maskGeoCoordinates(device.getGeo())));
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java:295]: The `PrivacyAnonymizationService` deals with TCF, CCPA and COPPA privacies. Why it cares about geo?

**Review by rpanchyk — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    public static PrivacyDebugLog from(Privacy originPrivacy,
+                                       Privacy resolvedPrivacy,
+                                       TcfContext tcfContext,
+                                       List<String> errors) {
+
+        return PrivacyDebugLog.builder()
+                .originPrivacy(makeOriginPrivacyDebug(originPrivacy))
+                .resolvedPrivacy(makeResolvedPrivacyDebug(resolvedPrivacy, tcfContext))
+                .errors(errors)
+                .privacyActionsPerBidder(new HashMap<>())
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java:39]: Should we think, of immutable data structures?

File: src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    public void addBidderLog(String bidder, Set<String> logs) {
+        if (CollectionUtils.isNotEmpty(logs) && StringUtils.isNotBlank(bidder)) {
+            privacyActionsPerBidder.put(bidder, logs);
+        }
+    }
+
+    @Value(staticConstructor = "of")
+    private static class PrivacyDebug {
+
+        @JsonInclude(value = JsonInclude.Include.NON_EMPTY)
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java:84]: Maybe move it on class level?

File: src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java (hunk 2 above)
Comment (snahornyi) [inline src/main/java/org/prebid/server/privacy/model/PrivacyDebugLog.java:97]: Let's place it after field declaration

File: src/main/java/org/prebid/server/privacy/PrivacyExtractor.java [hunk 7]
```diff
@@ -41,8 +43,15 @@
      * </ul><p>
      * And construct {@link Privacy} from them. Use default values in case of invalid value.
      */
-    public Privacy validPrivacyFrom(BidRequest bidRequest, List<String> errors) {
-        return extractPrivacy(bidRequest.getRegs(), bidRequest.getUser(), errors);
+    public PrivacyExtractionResult validPrivacyFrom(BidRequest bidRequest) {
+        final List<String> errors = new ArrayList<>();
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/privacy/PrivacyExtractor.java:47]: immutable collection?

File: src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                aliases.resolveBidder(bidder),
+                requestType,
+                userIdRemoved,
+                geoMasked,
+                analyticsBlocked,
+                requestBlocked);
+    }
+
+    /**
+     * Returns true if {@link User} has sensitive privacy information that can be masked.
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/privacy/PrivacyAnonymizationService.java:194]: can or should?

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 9]
```diff
@@ -526,14 +527,52 @@ private boolean isValidBidder(String bidder, BidderAliases aliases) {
...
+            Map<String, ExtBidderConfigOrtb> biddersToConfigs,
+            BidderAliases aliases) {
+
+        final BidRequest bidRequest = auctionContext.getBidRequest();
+        final Map<String, JsonNode> bidderToPrebidBidders = bidderToPrebidBidders(bidRequest);
+
+        final List<AuctionParticipation> bidderRequests = bidderPrivacyResults.stream()
+                .map(bidderPrivacyResult -> updatePrivacyDebugLog(
+                        bidderPrivacyResult, auctionContext.getPrivacyContext().getPrivacyDebugLog()))
+                // for each bidder create a new request that is a copy of original request except buyerid, imp
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/auction/ExchangeService.java:556]: Let's delete or move to method docs

**Review by snahornyi — APPROVED** [review]

**Pull request closed without merging**